### PR TITLE
feat: Create a GET endpoint to Retrieve Client Testimonials

### PR DESCRIPTION
**Pull request by chionz**
## Description

Created an API GET endpoint `/api/testimonials` that allows users to retrieve client testimonials. The endpoint supports fetching all testimonials with pagination, includes testimonial metadata, requires user authentication, and has proper error handling and rate limiting.

## Related Issue (Link to issue ticket)
[FEAT] Create a GET endpoint to Retrieve Client Testimonials #104

## Motivation and Context

This change is required to provide users with a way to easily access and retrieve client testimonials from our database. The endpoint is designed to handle large numbers of testimonials efficiently with pagination and includes necessary metadata such as client name, designation, date, and rating. It also implements rate limiting to restrict the number of requests a user can make to the endpoint.

## How Has This Been Tested?

- Added exactly 6 test testimonials to the database using `“seed.py”` and sent a GET request to `/api/testimonials?page=1` to verify correct pagination.
- Added a large number of testimonials (1000) to the database and sent a GET request to `/api/testimonials?page=1` to test handling of maximum allowable testimonials.
- Sent GET requests with invalid `rating` and `date` parameters to ensure proper validation and error handling.
- Implemented unit tests to test fetching all testimonials, pagination functionality, and error handling for invalid requests.

## Screenshots (if appropriate - Postman, etc):

Screenshot from 2024-07-21 19-18-52
Screenshot from 2024-07-22 10-25-50
Screenshot from 2024-07-22 10-23-57
Screenshot from 2024-07-22 10-24-43
Screenshot from 2024-07-22 10-25-29

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by Goketech — CHANGES_REQUESTED** [review]
17 files changed for this pr.

Pull changes made to dev branch, work with new models created.

Don't tamper with the migrations

File: api/utils/auth.py [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
+from jose import JWTError, jwt
 import bcrypt
 from api.v1.schemas.token import TokenData
 from api.db.database import get_db
 from .config import SECRET_KEY, ALGORITHM
 
-# Initialize OAuth2PasswordBearer
-oauth2_scheme = OAuth2PasswordBearer(tokenUrl="login")
+# Initialize HTTPBearer
+bearer_scheme = HTTPBearer()
```
Comment (joboy-dev) [inline api/utils/auth.py:16]: Please, what is this??

File: api/utils/auth.py [hunk 2]
```diff
@@ -43,4 +42,24 @@ def create_access_token(data: dict, expires_delta: Optional[timedelta] = None):
         expire = datetime.utcnow() + timedelta(minutes=15)
     to_encode.update({"exp": expire})
     encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
-    return encoded_jwt
\ No newline at end of file
+    return encoded_jwt
+
+def get_current_user(db: Session = Depends(get_db), credentials: HTTPAuthorizationCredentials = Security(bearer_scheme)):
```
Comment (joboy-dev) [inline api/utils/auth.py:47]: What is this for??
We have this already

File: api/v1/models/testimonials.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        DateTime,
+        func,
+        )
+from datetime import datetime
+from api.v1.models.base import Base
+from api.v1.models.base_model import BaseModel
+from sqlalchemy.dialects.postgresql import UUID
+
+
+class Testimonial(BaseModel, Base):
```
Comment (joboy-dev) [inline api/v1/models/testimonials.py:19]: Remove Base from here

File: api/v1/models/testimonials.py
Comment (joboy-dev) [inline api/v1/models/testimonials.py:1]: No need for id, created_at, updated_at

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Make changes

**Comment by chionz** [comment]
@joboy-dev I will close this PR. The new PR is at https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/pull/481
check it out please

**Comment by chionz** [comment]
@joboy-dev I will close this PR. The new PR is at https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/pull/481
check it out please

**Pull request closed without merging**